src/sentry/models/debugfile.py

```diff
@@ -497,212 +628,242 @@ def cache_path(self):
+        file.putfile(cache.open_stream())
+
+        # Try to insert the new Cache into the database. This only fail if
+        # (1) another process has concurrently added the same sym cache, or if
+        # (2) the debug symbol was deleted, either due to a newer upload or via
+        # the API.
+        try:
+            with transaction.atomic():
+                return cls.objects.create(
+                    project=debug_file.project,
+                    cache_file=file,
+                    debug_file=debug_file,
+                    checksum=debug_file.file.checksum,
+                    version=cache.file_format_version,
+                ), cache, None
+        except IntegrityError:
+            file.delete()
+
+        # Check for a concurrently inserted cache and use that instead. This
+        # could have happened (1) due to a concurrent insert, or (2) a new
+        # upload that has already succeeded to compute a cache. The latter
+        # case is extremely unlikely.
+        cache_file = cls.objects \
+            .filter(project=debug_file.project, debug_file__debug_id=debug_id) \
+            .select_related('cache_file') \
+            .order_by('-id') \
+            .first()
+
+        if cache_file is not None:
+            return cache_file, None, None
+
+        # There was no new cache, indicating that the debug file has been
+        # replaced with a newer version. Another job will create the
+        # corresponding cache eventually. To prevent querying the database
+        # another time, simply use the in-memory cache for now:
+        return None, cache, None
+
+    def _load_cachefiles_via_fs(self, project, cachefiles, cls):
         rv = {}
         base = self.get_project_path(project)
-        for debug_id, symcache_file in cachefiles:
-            cachefile_path = os.path.join(base, debug_id + '.symcache')
+        cls_name = cls.__name__.lower()
+
+        for debug_id, model, cache in cachefiles:
+            # If we're given a cache instance, use that over accessing the file
+            # system or potentially even blob storage.
+            if cache is not None:
+                rv[debug_id] = cache
+                continue
+            elif model is None:
+                raise RuntimeError('missing %s file to load from fs' % cls_name)
+
+            # Try to locate a cached instance from the file system and bump the
+            # timestamp to indicate it is still being used. Otherwise, download
+            # from the blob store and place it in the cache folder.
+            cachefile_name = '%s_%s.%s' % (model.id, model.version, cls_name)
+            cachefile_path = os.path.join(base, cachefile_name)
             try:
                 stat = os.stat(cachefile_path)
             except OSError as e:
                 if e.errno != errno.ENOENT:
                     raise
-                symcache_file.cache_file.save_to(cachefile_path)
+                model.cache_file.save_to(cachefile_path)
```

Comment:
noob question - How do local cache files get removed?

Reply:
There's a periodically scheduled job `clean_dsymcache` that deletes files from the FS cache when they haven't been referenced for 1.5 days or longer. The actual "logic" is defined in this file at the very bottom: `clear_old_entries`.

Reply:
Thank you 🙇